src/http/common.cc

```diff
@@ -134,6 +134,48 @@ output_stream<char> make_http_chunked_output_stream(output_stream<char>& out) {
     return output_stream<char>(http_chunked_data_sink(out), default_body_sink_buffer_size, opts);
 }
 
+class http_content_length_data_sink_impl : public data_sink_impl {
+    output_stream<char>& _out;
+    const size_t _limit;
+    size_t& _bytes_written;
+
+public:
+    http_content_length_data_sink_impl(output_stream<char>& out, size_t& len)
+            : _out(out)
+            , _limit(std::exchange(len, 0))
+            , _bytes_written(len)
+    {
+    }
+    virtual future<> put(net::packet data)  override { abort(); }
+    using data_sink_impl::put;
+    virtual future<> put(temporary_buffer<char> buf) override {
+        if (buf.size() == 0 || _bytes_written == _limit) {
+            return make_ready_future<>();
+        }
+        auto size = std::min(buf.size(), _limit - _bytes_written);
+        return _out.write(buf.get(), size).then([this, size] {
+            _bytes_written += size;
+        });
+    }
+    virtual future<> close() override {
+        return make_ready_future<>();
+    }
+};
+
+class http_content_length_data_sink : public data_sink {
+public:
+    http_content_length_data_sink(output_stream<char>& out, size_t& len)
+        : data_sink(std::make_unique<http_content_length_data_sink_impl>(out, len))
+    {
+    }
+};
+
+output_stream<char> make_http_content_length_output_stream(output_stream<char>& out, size_t& len) {
+    output_stream_options opts;
+    opts.trim_to_size = true;
```

Comment:
What does this trim_to_size thing do?

Reply:
It makes sure the stream that sits on top of the sink doesn't supply buffers larger than the configured one.

I copied it from make_http_chunked_output_stream() which's there since 8eed9671dbd7a2f886f4970afb356bc4843da307 without explanation why trim-to-size matters

Reply:
Actually, I think that you added this trim_to_size=true thing in commit 66202917709562ceafb18fa664d95b2b1e6b52f3 - you added it in make_http_chunked_output_stream and in other places.
I still don't understand what benefits this option achieves, and what happens if we don't use it.
I saw that @avikivity asked you to write a test of sending larger than 32,000 bytes (where did this weird number come from, by the way?) to exercise this trim_to_size thing. I'm still puzzled what it does and why it's needed.

Reply:
> Actually, I think that you added this trim_to_size=true thing in commit https://github.com/scylladb/seastar/commit/66202917709562ceafb18fa664d95b2b1e6b52f3 - you added it in make_http_chunked_output_stream and in other places.

No I didn't. I deprecated the `output_stream(data_sink fd, size_t size, bool trim_to_size, bool batch_flushes = false)` and made all its callers of `output_stream(fd, size, true)` kind to use `opts.trim_to_size instead`

Reply:
Oh, you're right. I missed that innocent-looking "true" :-(

I really don't understand where these 32000 and trim-to-size come from. It even looks like a mistake to me - there is no reason why chunked encoding can't have one chunk of size 1KB and then one chunk of size 1MB if the code has 1MB it wants to send together. Chunks don't have to be small! But I understand this is not a new thing, and we don't need to consider it or fix it now. Ok.